### PR TITLE
menu-ui/ENG-3744 implement menu mfe

### PR DESCRIPTION
**Pull request by ichalagashvili**
PENDING:

- There are some unclear or to be done items marked with @TODO in the code

### CODE REVIEW

File: microfrontends/app-builder-menu/mocks/handlers.ts [hunk 1]
```diff
@@ -1,13 +1,46 @@
...
   rest.get('/your-endpoint', (req, res, ctx) => {
     return res(
       ctx.status(200),
       ctx.json({
-        message: "hello world"
+        message: 'hello world'
       })
-    )
+    );
   }),
```
Comment (Dv92) [inline microfrontends/app-builder-menu/mocks/handlers.ts]: please delete this example

File: microfrontends/app-builder-menu/src/custom-element.tsx [hunk 2]
```diff
@@ -13,6 +14,41 @@ export class AppBuilderMenu extends HTMLElement {
...
+      'manageCategories',
+      'validateContents',
+      'manageResources',
+      'enterECR'
+    ],
+    lang: 'en',
+    api: {
+      url: 'http://localhost:8080/menu-be-api'
+    }
+  };
```
Comment (Dv92) [inline microfrontends/app-builder-menu/src/custom-element.tsx]: move this mocks in the index.html

File: microfrontends/app-builder-menu/package.json [hunk 3]
```diff
@@ -12,6 +12,7 @@
     "axios": "0.27.2",
     "react": "18.0.0",
     "react-dom": "18.0.0",
+    "react-router-dom": "6.3.0",
```
Comment (Dv92) [inline microfrontends/app-builder-menu/package.json]: please use window.appBuilderRouter for the navigation. It's the method for navigate the MFE inside AB.

File: microfrontends/app-builder-menu/src/utils/dynamicTree.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+): DYNAMIC_MENU_ITEM[] => {
+  const uniquePbcs = [...new Set(items.map(item => item.pbcName))];
+  return uniquePbcs.map(pbc => ({
+    parent: pbc,
+    children: items
+      .filter(item => item.pbcName === pbc)
+      .map(item => ({
+        ...item,
+        ...(item.target === EXTERNAL_MFE_TARGET_NAME
+          ? { externalHref: item.addr }
```
Comment (Dv92) [inline microfrontends/app-builder-menu/src/utils/dynamicTree.ts]: maybe here you can simplify the contract in this way `{ url: item.addr, target: "_blank", ref: "nopener noreferrer" }`

**Review by Dv92 — COMMENTED** [review]
- prefer the named exports over default exports https://basarat.gitbook.io/typescript/main-1/defaultisbad
- please add the tests
- for the presentational components add the snapshot testing https://jestjs.io/docs/snapshot-testing

File: microfrontends/app-builder-menu/src/components/Menu.tsx [hunk 5]
```diff
@@ -1,33 +1,94 @@
...
 
-const MenuCmp = styled.menu`
-  position: fixed;
-  left: 0;
-  top: 0;
-  background-color: black;
-  width: 250px;
-  height: 100%;
-  color: white;
+const GlobalStyle = createGlobalStyle`
```
Comment (gReis89) [inline microfrontends/app-builder-menu/src/components/Menu.tsx]: what if we move it to a `styles.ts` file?

File: microfrontends/app-builder-menu/src/types/api.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
+export interface MENU_ITEM {
```
Comment (gReis89) [inline microfrontends/app-builder-menu/src/types/api.ts]: I'm used to use PascalCase for types and interfaces to avoid using `_` what do u think @Dv92

Comment (Dv92) [reply microfrontends/app-builder-menu/src/types/api.ts]: I agree, usually, the notation uppercase snake is for `const`. https://basarat.gitbook.io/typescript/styleguide#interface

Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/types/api.ts]: Fixed

File: microfrontends/app-builder-menu/src/components/Menu.tsx (hunk 5 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/components/Menu.tsx]: GlobalStyled will be moved in the end, so does not make sense to move it now.

File: microfrontends/app-builder-menu/src/utils/dynamicTree.ts (hunk 4 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/utils/dynamicTree.ts]: Fixed

File: microfrontends/app-builder-menu/package.json (hunk 3 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/package.json]: Fixed

File: microfrontends/app-builder-menu/src/custom-element.tsx (hunk 2 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/custom-element.tsx]: Fixed

File: microfrontends/app-builder-menu/mocks/handlers.ts (hunk 1 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/mocks/handlers.ts]: Fixed

File: microfrontends/app-builder-menu/src/utils/navigation.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export const useNavigation = () => {
+  return (route: string) => window.appBuilderRouter.push(route);
+};
```
Comment (Dv92) [inline microfrontends/app-builder-menu/src/utils/navigation.ts]: @ichalagashvili @gReis89 We could move this hook into a separate directory called `hooks`, in this way we can filter `utils` from `react hooks`. What do you think?

- `hooks`: contain all react custom hooks related to the project.
- `utils`: contain all the utils functions unrelated to react.

File: microfrontends/app-builder-menu/src/components/MenuUI.tsx [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+            <ListGroupItem
+              id="pages"
+              label="Pages"
+              className="app-tour-step-3"
+              renderIcon={props => <PagesIcon {...props} />}
+              onClick={() => {
+                const event = new CustomEvent('tutorial', {
+                  detail: { nextStep: 4 }
+                });
+                window.dispatchEvent(event);
```
Comment (Dv92) [inline microfrontends/app-builder-menu/src/components/MenuUI.tsx]: I see that this code is duplicated in other handlers after this. I suggest isolating this logic inside a function and reuse inside the other handlers.

File: microfrontends/app-builder-menu/src/components/Menu.tsx [hunk 9]
```diff
@@ -1,33 +1,93 @@
...
+}
+table {
+	border-collapse: collapse;
+	border-spacing: 0;
+}
+
+* {
+  box-sizing: border-box;
+}
 `;
```
Comment (Dv92) [inline microfrontends/app-builder-menu/src/components/Menu.tsx]: can you please delete this global styles? You have moved in the index.html so you didn't need anymore

File: microfrontends/app-builder-menu/src/components/ListGroup/ListGroup.test.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+//       onClick={() => navigate(ROUTE_PAGE_TEMPLATE_LIST)}
+//     />
+//     <SecondaryMenuItem
+//       id="pages-settings"
+//       label="Settings"
+//       onClick={() => navigate(ROUTE_PAGE_SETTINGS)}
+//     />
+//   </>
+// )}
+// </ListGroupItem>
```
Comment (Dv92) [inline microfrontends/app-builder-menu/src/components/ListGroup/ListGroup.test.tsx]: can you delete this commented code?

Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/components/ListGroup/ListGroup.test.tsx]: done

File: microfrontends/app-builder-menu/src/components/MenuUI.tsx (hunk 8 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/components/MenuUI.tsx]: done

File: microfrontends/app-builder-menu/src/utils/navigation.ts (hunk 7 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/utils/navigation.ts]: done

File: microfrontends/app-builder-menu/src/components/Menu.tsx (hunk 9 above)
Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/components/Menu.tsx]: done ✅

File: microfrontends/app-builder-menu/src/types/api.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    items: MenuItem[];
+  };
+}
+
+export interface EpcData {
+  'data-organization'?: string; // @TODO Marco tell me final answer after talk with Walter
+  'data-epc-id'?: string; // pbcName
+  'data-submenu'?: string; // @TODO Marco tell me final answer after talk with Nicola
+  'data-menu-item-id'?: string; // we generate this
+}
```
Comment (Dv92) [inline microfrontends/app-builder-menu/src/types/api.ts]: @ichalagashvili for this one:
- data-organization: for now add a field `organization` on the menu items and use that.
- data-epc-id: is the pbcName on the API response
- submenu: populated only for parent voices that have some children (for example "strapi"). Use pbcName
- data-menu-item-id: please use in this case pbcName + bundleName + label en (as snakecase)

Comment (ichalagashvili) [reply microfrontends/app-builder-menu/src/types/api.ts]: done

**Review by Dv92 — APPROVED** [review]

**Pull request merged**